From 695bf1bb541efcbe46cf842bf51fdd35b9d6a5af Mon Sep 17 00:00:00 2001
From: Ken Tseng <ken.tseng@realtek.com>
Date: Wed, 22 Apr 2026 15:45:03 +0800
Subject: [PATCH] spl: add BOOT_DEVICE_SPI_NAND for SPL_SPI_NAND_LOAD

- Try second copy on second SPL_SPI_NAND_LOAD
- mtd: SYS_NAND_MAX_CHIPS: add MTD_SPI_NAND depends
- spl_ubi: support BOOT_DEVICE_NAND on SPL_SPI_NAND_LOAD
-- return spl_parse_image_header status for next boot order
---
 arch/arm/include/asm/spl.h |  1 +
 common/spl/spl_spi_nand.c  |  8 +++++---
 common/spl/spl_ubi.c       | 30 ++++++++++++++++++++++++++++++
 drivers/mtd/Kconfig        |  2 +-
 4 files changed, 37 insertions(+), 4 deletions(-)

diff --git a/arch/arm/include/asm/spl.h b/arch/arm/include/asm/spl.h
index ee79a19c05..273c64007a 100644
--- a/arch/arm/include/asm/spl.h
+++ b/arch/arm/include/asm/spl.h
@@ -30,6 +30,7 @@ enum {
 	BOOT_DEVICE_XIP,
 	BOOT_DEVICE_BOOTROM,
 	BOOT_DEVICE_SMH,
+	BOOT_DEVICE_SPI_NAND,
 	BOOT_DEVICE_NONE
 };
 #endif
diff --git a/common/spl/spl_spi_nand.c b/common/spl/spl_spi_nand.c
index 4e37ad91b3..a9aaf3fc4b 100644
--- a/common/spl/spl_spi_nand.c
+++ b/common/spl/spl_spi_nand.c
@@ -25,7 +25,7 @@
 #include <dm/uclass.h>
 #include <mtd.h>
 
-static struct mtd_info *spl_spi_nand_get_mtd(void)
+struct mtd_info *spl_spi_nand_get_mtd(void)
 {
 	struct udevice *dev;
 	int ret;
@@ -66,6 +66,8 @@ static int spl_spinand_load_image(struct spl_image_info *spl_image,
 {
 	struct spl_load_info load;
 	struct mtd_info *mtd;
+	static int time = 0;
+	size_t offset = CONFIG_SYS_SPI_U_BOOT_OFFS + 0xC0000 * (time++);
 
 	mtd = spl_spi_nand_get_mtd();
 	if (!mtd) {
@@ -79,8 +81,8 @@ static int spl_spinand_load_image(struct spl_image_info *spl_image,
 	load.read = spl_spinand_fit_read;
 	load.priv = mtd;
 
-	return spl_load(spl_image, bootdev, &load, 0, CONFIG_SYS_SPI_U_BOOT_OFFS);
+	return spl_load(spl_image, bootdev, &load, 0, offset);
 }
 
 /* Use priority 1 so that boards can override this */
-SPL_LOAD_IMAGE_METHOD("SPI NAND", 1, BOOT_DEVICE_SPI, spl_spinand_load_image);
+SPL_LOAD_IMAGE_METHOD("SPI NAND", 1, BOOT_DEVICE_SPI_NAND, spl_spinand_load_image);
diff --git a/common/spl/spl_ubi.c b/common/spl/spl_ubi.c
index d7ab9efd11..a5cd4f97e5 100644
--- a/common/spl/spl_ubi.c
+++ b/common/spl/spl_ubi.c
@@ -11,6 +11,25 @@
 #include <onenand_uboot.h>
 #include <ubispl.h>
 #include <spl.h>
+#ifdef CONFIG_SPL_SPI_NAND_LOAD
+#include <mtd.h>
+
+static struct mtd_info *mtd = NULL;
+
+struct mtd_info *spl_spi_nand_get_mtd(void);
+static int spi_nand_spl_read_block(int block, int offset, int len, void *dst) {
+	loff_t offs = block * mtd->erasesize + offset;
+	size_t retlen = 0;
+	int ret;
+
+	ret = mtd_read(mtd, offs, len, &retlen, dst);
+	if (ret && ret != -EUCLEAN) {
+		printf("SPI NAND read failed offs=0x%lx len=0x%lx ret=%d\n",
+		       offs, len, ret);
+	}
+	return 0;
+}
+#endif
 
 int spl_ubi_load_image(struct spl_image_info *spl_image,
 		       struct spl_boot_device *bootdev)
@@ -33,6 +52,17 @@ int spl_ubi_load_image(struct spl_image_info *spl_image,
 		info.read = onenand_spl_read_block;
 		info.peb_size = CFG_SYS_ONENAND_BLOCK_SIZE;
 		break;
+#endif
+#ifdef CONFIG_SPL_SPI_NAND_LOAD
+	case BOOT_DEVICE_NAND:
+		mtd = spl_spi_nand_get_mtd();
+		if (!mtd) {
+			puts("SPI NAND probe for UBI failed.\n");
+			return -ENODEV;
+		}
+		info.read = spi_nand_spl_read_block;
+		info.peb_size = mtd->erasesize;
+		break;
 #endif
 	default:
 		goto out;
@@ -75,7 +105,7 @@ int spl_ubi_load_image(struct spl_image_
 
 	ret = ubispl_load_volumes(&info, volumes, 1);
 	if (!ret)
-		spl_parse_image_header(spl_image, bootdev, header);
+		ret = spl_parse_image_header(spl_image, bootdev, header);
 out:
 #ifdef CONFIG_SPL_NAND_SUPPORT
 	if (bootdev->boot_device == BOOT_DEVICE_NAND)
diff --git a/drivers/mtd/Kconfig b/drivers/mtd/Kconfig
index a9a32fc2c2..2bfb17ab99 100644
--- a/drivers/mtd/Kconfig
+++ b/drivers/mtd/Kconfig
@@ -270,7 +270,7 @@ config SYS_NAND_MAX_ECCPOS
 config SYS_NAND_MAX_CHIPS
 	int "NAND max chips"
 	depends on MTD_RAW_NAND || CMD_ONENAND || TARGET_S5PC210_UNIVERSAL || \
-		SPL_OMAP3_ID_NAND
+		SPL_OMAP3_ID_NAND || MTD_SPI_NAND
 	default 1
 	help
 	  The maximum number of NAND chips per device to be supported.
-- 
2.47.0

